### ISSUE TITLE
[Issue]: Reinstall force can cause corrupt VENV

### ISSUE THREAD

**Issue by TKS4** [vladmandic/automatic#868]
### Issue Description

With a recent error that was patched hastily https://github.com/vladmandic/automatic/issues/865 there seems to be an error with basicsr upgrade argument similarly.

`2023-05-11 11:31:07,743 | DEBUG | E:\Stable-Diffusion-vlad\automatic\installer.py | Package version found: basicsr 1.4.2
2023-05-11 11:31:07,743 | INFO | E:\Stable-Diffusion-vlad\automatic\installer.py | Installing package: basicsr
2023-05-11 11:31:07,743 | DEBUG | E:\Stable-Diffusion-vlad\automatic\installer.py | Running pip: install --upgrade --force basicsr
2023-05-11 11:32:21,126 | ERROR | E:\Stable-Diffusion-vlad\automatic\installer.py | Error running pip: install --upgrade --force basicsr
2023-05-11 11:32:21,141 | DEBUG | E:\Stable-Diffusion-vlad\automatic\installer.py | Pip output: Collecting basicsr

  Using cached basicsr-1.4.2-py3-none-any.whl

Collecting addict (from basicsr)

  Using cached addict-2.4.0-py3-none-any.whl (3.8 kB)

Collecting future (from basicsr)

  Using cached future-0.18.3-py3-none-any.whl

Collecting lmdb (from basicsr)

  Using cached lmdb-1.4.1-cp310-cp310-win_amd64.whl (100 kB)

Collecting numpy>=1.17 (from basicsr)

  Using cached numpy-1.24.3-cp310-cp310-win_amd64.whl (14.8 MB)

Collecting opencv-python (from basicsr)

  Using cached opencv_python-4.7.0.72-cp37-abi3-win_amd64.whl (38.2 MB)

Collecting Pillow (from basicsr)

  Using cached Pillow-9.5.0-cp310-cp310-win_amd64.whl (2.5 MB)

Collecting pyyaml (from basicsr)

  Using cached PyYAML-6.0-cp310-cp310-win_amd64.whl (151 kB)

Collecting requests (from basicsr)

  Using cached requests-2.30.0-py3-none-any.whl (62 kB)

Collecting scikit-image (from basicsr)

  Using cached scikit_image-0.20.0-cp310-cp310-win_amd64.whl (23.7 MB)

Collecting scipy (from basicsr)

  Using cached scipy-1.10.1-cp310-cp310-win_amd64.whl (42.5 MB)

Collecting tb-nightly (from basicsr)

  Downloading tb_nightly-2.14.0a20230511-py3-none-any.whl (5.6 MB)

     ---------------------------------------- 5.6/5.6 MB 13.7 MB/s eta 0:00:00

Collecting torch>=1.7 (from basicsr)

  Downloading torch-2.0.1-cp310-cp310-win_amd64.whl (172.3 MB)

     -------------------------------------- 172.3/172.3 MB 8.4 MB/s eta 0:00:00

Collecting torchvision (from basicsr)

  Downloading torchvision-0.15.2-cp310-cp310-win_amd64.whl (1.2 MB)

     ---------------------------------------- 1.2/1.2 MB 10.8 MB/s eta 0:00:00

Collecting tqdm (from basicsr)

  Using cached tqdm-4.65.0-py3-none-any.whl (77 kB)

Collecting yapf (from basicsr)

  Using cached yapf-0.33.0-py2.py3-none-any.whl (200 kB)

Collecting filelock (from torch>=1.7->basicsr)

  Using cached filelock-3.12.0-py3-none-any.whl (10 kB)

Collecting typing-extensions (from torch>=1.7->basicsr)

  Using cached typing_extensions-4.5.0-py3-none-any.whl (27 kB)

Collecting sympy (from torch>=1.7->basicsr)

  Downloading sympy-1.12-py3-none-any.whl (5.7 MB)

     ---------------------------------------- 5.7/5.7 MB 13.1 MB/s eta 0:00:00

Collecting networkx (from torch>=1.7->basicsr)

  Using cached networkx-3.1-py3-none-any.whl (2.1 MB)

Collecting jinja2 (from torch>=1.7->basicsr)

  Using cached Jinja2-3.1.2-py3-none-any.whl (133 kB)

Collecting charset-normalizer<4,>=2 (from requests->basicsr)

  Using cached charset_normalizer-3.1.0-cp310-cp310-win_amd64.whl (97 kB)

Collecting idna<4,>=2.5 (from requests->basicsr)

  Using cached idna-3.4-py3-none-any.whl (61 kB)

Collecting urllib3<3,>=1.21.1 (from requests->basicsr)

  Using cached urllib3-2.0.2-py3-none-any.whl (123 kB)

Collecting certifi>=2017.4.17 (from requests->basicsr)

  Using cached certifi-2023.5.7-py3-none-any.whl (156 kB)

Collecting imageio>=2.4.1 (from scikit-image->basicsr)

  Using cached imageio-2.28.1-py3-none-any.whl (3.4 MB)

Collecting tifffile>=2019.7.26 (from scikit-image->basicsr)

  Using cached tifffile-2023.4.12-py3-none-any.whl (219 kB)

Collecting PyWavelets>=1.1.1 (from scikit-image->basicsr)

  Using cached PyWavelets-1.4.1-cp310-cp310-win_amd64.whl (4.2 MB)

Collecting packaging>=20.0 (from scikit-image->basicsr)

  Using cached packaging-23.1-py3-none-any.whl (48 kB)

Collecting lazy_loader>=0.1 (from scikit-image->basicsr)

  Using cached lazy_loader-0.2-py3-none-any.whl (8.6 kB)

Collecting absl-py>=0.4 (from tb-nightly->basicsr)

  Using cached absl_py-1.4.0-py3-none-any.whl (126 kB)

Collecting grpcio>=1.48.2 (from tb-nightly->basicsr)

  Using cached grpcio-1.54.0-cp310-cp310-win_amd64.whl (4.1 MB)

Collecting google-auth<3,>=1.6.3 (from tb-nightly->basicsr)

  Using cached google_auth-2.18.0-py2.py3-none-any.whl (178 kB)

Collecting google-auth-oauthlib<1.1,>=0.5 (from tb-nightly->basicsr)

  Using cached google_auth_oauthlib-1.0.0-py2.py3-none-any.whl (18 kB)

Collecting markdown>=2.6.8 (from tb-nightly->basicsr)

  Using cached Markdown-3.4.3-py3-none-any.whl (93 kB)

Collecting protobuf>=3.19.6 (from tb-nightly->basicsr)

  Using cached protobuf-4.23.0-cp310-abi3-win_amd64.whl (422 kB)

Collecting setuptools>=41.0.0 (from tb-nightly->basicsr)

  Using cached setuptools-67.7.2-py3-none-any.whl (1.1 MB)

Collecting tensorboard-data-server<0.8.0,>=0.7.0 (from tb-nightly->basicsr)

  Using cached tensorboard_data_server-0.7.0-py3-none-any.whl (2.4 kB)

Collecting werkzeug>=1.0.1 (from tb-nightly->basicsr)

  Using cached Werkzeug-2.3.4-py3-none-any.whl (242 kB)

Collecting wheel>=0.26 (from tb-nightly->basicsr)

  Using cached wheel-0.40.0-py3-none-any.whl (64 kB)

Collecting colorama (from tqdm->basicsr)

  Using cached colorama-0.4.6-py2.py3-none-any.whl (25 kB)

Collecting tomli>=2.0.1 (from yapf->basicsr)

  Using cached tomli-2.0.1-py3-none-any.whl (12 kB)

Collecting cachetools<6.0,>=2.0.0 (from google-auth<3,>=1.6.3->tb-nightly->basicsr)

  Using cached cachetools-5.3.0-py3-none-any.whl (9.3 kB)

Collecting pyasn1-modules>=0.2.1 (from google-auth<3,>=1.6.3->tb-nightly->basicsr)

  Using cached pyasn1_modules-0.3.0-py2.py3-none-any.whl (181 kB)

Collecting six>=1.9.0 (from google-auth<3,>=1.6.3->tb-nightly->basicsr)

  Using cached six-1.16.0-py2.py3-none-any.whl (11 kB)

Collecting urllib3<3,>=1.21.1 (from requests->basicsr)

  Using cached urllib3-1.26.15-py2.py3-none-any.whl (140 kB)

Collecting rsa<5,>=3.1.4 (from google-auth<3,>=1.6.3->tb-nightly->basicsr)

  Using cached rsa-4.9-py3-none-any.whl (34 kB)

Collecting requests-oauthlib>=0.7.0 (from google-auth-oauthlib<1.1,>=0.5->tb-nightly->basicsr)

  Using cached requests_oauthlib-1.3.1-py2.py3-none-any.whl (23 kB)

Collecting MarkupSafe>=2.1.1 (from werkzeug>=1.0.1->tb-nightly->basicsr)

  Using cached MarkupSafe-2.1.2-cp310-cp310-win_amd64.whl (16 kB)

Collecting mpmath>=0.19 (from sympy->torch>=1.7->basicsr)

  Downloading mpmath-1.3.0-py3-none-any.whl (536 kB)

     ------------------------------------- 536.2/536.2 kB 17.0 MB/s eta 0:00:00

Collecting pyasn1<0.6.0,>=0.4.6 (from pyasn1-modules>=0.2.1->google-auth<3,>=1.6.3->tb-nightly->basicsr)

  Using cached pyasn1-0.5.0-py2.py3-none-any.whl (83 kB)

Collecting oauthlib>=3.0.0 (from requests-oauthlib>=0.7.0->google-auth-oauthlib<1.1,>=0.5->tb-nightly->basicsr)

  Using cached oauthlib-3.2.2-py3-none-any.whl (151 kB)

Installing collected packages: mpmath, lmdb, addict, wheel, urllib3, typing-extensions, tomli, tensorboard-data-server, sympy, six, setuptools, pyyaml, pyasn1, protobuf, Pillow, packaging, oauthlib, numpy, networkx, MarkupSafe, markdown, lazy_loader, idna, grpcio, future, filelock, colorama, charset-normalizer, certifi, cachetools, absl-py, yapf, werkzeug, tqdm, tifffile, scipy, rsa, requests, PyWavelets, pyasn1-modules, opencv-python, jinja2, imageio, torch, scikit-image, requests-oauthlib, google-auth, torchvision, google-auth-oauthlib, tb-nightly, basicsr
`

### Version Platform Description

Version: 44be0153

**Comment by vladmandic** [comment]
i see error reported on top that it failed, but not actuall error message?

**Comment by TKS4** [comment]
I'll need to run the --reinstall operation again, I'll update you with the full log in that case.

**Comment by vladmandic** [comment]
if you do, delete (if already exists) `setup.log` and upload it as file, much better than copy&paste.

**Comment by TKS4** [comment]
Okay, so I need to explain something as there is now a bigger problem as a result of using --reinstall. I will provide two setup.log's so you can see the chain of the two errors.

The things I did, nuked the venv directory, started webUI to repopulate it pretty quickly, then ran --reinstall (which takes a long time obviously), and then ran webUI again after the prior --reinstall process and found another two problems as a result (with the only way of fixing them, is by nuking the venv directory again).

The first problem is the reason I opened this issue. The actual error seems to be some sort of permissions error (the thing I forgot to include when I copy/pasted the log verbosity I thought was relevant. This is the precise error I think you were looking for: 

```
ERROR: Could not install packages due to an OSError: [WinError 5] Access is denied: 'E:\\Stable-Diffusion-vlad\\automatic\\venv\\Lib\\site-packages\\~vfuser\\_C.cp310-win_amd64.pyd'

Check the permissions.
```
This error log can be found in the first setup.log I will provide.

---------------------------------------------------------------

The 2 new issues arrise when I try to start webUI again (actually a big problem for someone like me who doesn't know how to manually fix it). The first major issue, is webUI for some reason doesn't want to use torch+cu118, instead it will now always fire up torch+cpu (meaning even though it detects the CUDA toolkit and all that, it simply defaults to loading up on the CPU. The second minor issue seems to be a result of the prior --reinstall we do. In that the resinstalling of packages lead to installing a higher version of urllib3:
```

Package wrong version: urllib3 2.0.2 required 1.26.15
13:15:53-809699 INFO     Installing package: urllib3==1.26.15
```

Both of these errors are logged in the second setup.log I will provide.

The urllbib3 gets downgraded, and the issue is not seen again. So I'm wondering if --reinstall is needlessly installing a more up to date version than WebUI wants, or if WebUI is needlessly downgrading to an older version for stability that you yourself explicitly want because V2.0.2

----------------------------------------------------------------

So to summarize, fresh venv generation > run --reinstall > basicsr install error > fire up webUI again > urllib downgrade occurs + webUI will only run on CPU > nuke venv > fresh venv generation leads to using GPU again

[2ndErrorLogg.log](https://github.com/vladmandic/automatic/files/11454983/2ndErrorLogg.log)
[reinstallLOG.log](https://github.com/vladmandic/automatic/files/11454984/reinstallLOG.log)

**Comment by vladmandic** [comment]
the problem with reinstall is that it **forces** reinstall of everything regardless of state - and for each package, it doesn't only install package, but all packages that package lists as its requirements. And those recursive requirements are often in conflict in python world since python keeps all libs in a flat format (unlike for example NodeJS where each requirements has its requirements in its own subfolder).

so it takes several restarts/installs after --reinstall until app can "fix" everything that force-installation of packages has done.

i've added --reinstall as debug/advanced use, looking-backwards, I should not not have had it publically announced at all.
I'll now remove `--force` flag so it doesn't install recursive conflicting packages.

On why torch picks up CPU version instead of CUDA? Whoever is doing release engineering at torch should be fired. They mess with binary wheels and detection regularly. One day its fine, the next its upside/down.
Yesterday `torchvision` was installing CUDA 11.7 version instead of CUDA 11.8 even if path was `https://download.pytorch.org/whl/cu118` simply because they put wrong binaries there.

You can run command manually (globally or inside venv):
`pip torch torchaudio torchvision==0.15.1 --index-url https://download.pytorch.org/whl/cu118`

And see if it installs correct torch.

**Comment by TKS4** [comment]
(I tried that command at the end of your post and I get an error: "unknown command torch"). This is with venv activated.

In terms of --reinstall doing all the things you said, tbh it's not a problem for me personally, I just thought maybe there was something wrong going on that you might want to know about yourself (since as I said, I can nuke the venv directory, and run webUI.bat normally, and everything gets repopulated pretty fast and runs without any real errors). I was just curious if basicsr was updating like you personally wanted it to, and also was wondering if urllib3 2.0.2 downgrading to 1.25.15 was also something you yourself wanted.

As far as the torch release engineering, LOL that was funny.

Btw there is one thing I am confused about if you don't mind me asking. What precisely would be the purpose of --resinstall as a function? I know you just told me it's for advanced use, but because I don't understand the logic (since I'm not a programmer), why would anyone want to do this given how long the process takes, when seemingly I can avoid any of these issues by simply nuking the venv folder, running webUI.bat, and then having it quickly repopulate whatever it needs seemingly without fail? Is it just to pull those aforementioned package requirements of packages so you know for sure nothing is missing whenever an update might occur? If it's too complex to explain to a layman, that's fine just skip this question. The reason I ask is because repopulating a nuked venv folder currently doesn't require all those packages of packages - so I'm wondering why it would matter to anyone besides contributors so they don't have to go manually hunting down for these sorts of things potentially when debugging.

**Comment by vladmandic** [comment]
> WARNING: Ignoring invalid distribution -orch (e:\stable-diffusion-vlad\automatic\venv\lib\site-packages)

That's python + venv + windows = can result in bad files. i'm amazed that python hasn't solved that.
How things typically work on linux is that anything new is created under temp name and then when everything is ready, its renamed. That's because on linux, rename is the file only operation that is guaranteed to be atomic (cannot be changed by some other process in parallel).

Windows on the other had creates locks on files in use, so there is no need for that. But it also means if you try to do quick create/rename, it may also fail because windows did not release lock just in time.

I've just fixed issue with images save failing sometimes on windows exactly because of that.
But i cannot fix it inside "pip" (python's installer).

**Comment by TKS4** [comment]
God Windows sucks.. A new USB stick I got will not unmount (Eject) if task manager is running if you can believe that. So much of that locking behavior with anything you do (sometimes I have to restart explorer.exe to get files unlocked).

Seeing as how Proton has been making strides in the last few months - slowly any reason I have for staying with this dumb OS is depleting.

**Comment by vladmandic** [comment]
Btw, Im closing the issue, but feel free to comment...
